Bluetooth: GATT: Reduce footprint of client parameters

Vudentz opened the pull request:
This reduces the overall footprint of GATT client parameters so they don't carry bt_att_req themselves, it also optimize bt_gatt_subscribe_params so they don't have to carry the peer address themselves thus reducing their footprint even further.

Vudentz committed:
Bluetooth: GATT: Allocate request from a memory slab

This should reduce the footprint on applications that do a lot of
requests i.e have a lot of subscriptions.

Fixes #21103

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Remove _peer member from bt_gatt_subscribe_params

This further reduce the overhead on each subscription at expense of
having a dedicated array to store subscriptions, the code now maintain
a separate list for each peer which should also scale better with large
number of subscriptions to different peers.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -61,6 +61,9 @@ NET_BUF_POOL_DEFINE(prep_pool, CONFIG_BT_ATT_PREPARE_COUNT, BT_ATT_MTU,
 #endif /* CONFIG_BT_ATT_PREPARE_COUNT */
 
+K_MEM_POOL_DEFINE(req_pool, sizeof(struct bt_att_req),
+		  sizeof(struct bt_att_req), CONFIG_BT_ATT_TX_MAX, 16);
```
Wouldn't `k_mem_slab` be the more appropriate data structure to use since you're allocating fixed same-sized blocks of data? Using `k_mem_pool` has the added cost of supporting arbitrary-sized allocations which you're not taking advantage of.

Vudentz replied:
Will change to k_mem_slab.

Vudentz replied:
Done.

jhedberg commented on subsys/bluetooth/host/att_internal.h:
```diff
@@ -6,6 +6,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#define BT_EATT_PSM		0x7f
```
Seems unrelated to this PR?

jhedberg commented on subsys/bluetooth/host/att_internal.h:
```diff
@@ -222,6 +224,26 @@ struct bt_att_signature {
+	u16_t len;
+	u8_t  value[0];
+} __packed;
+
```
These also seem unrelated?

Vudentz replied:
Fixed.

jhedberg commented on subsys/bluetooth/host/att_internal.h:
```diff
@@ -139,6 +140,7 @@ struct bt_att_read_mult_req {
 /* Read Multiple Respose */
 #define BT_ATT_OP_READ_MULT_RSP			0x0f
 struct bt_att_read_mult_rsp {
+	u16_t len;
```
Unrelated?

Vudentz replied:
Fixed.

jhedberg reviewed (changes requested):
These seems to be a bunch of unrelated changes that have slipped into this PR from your EATT work.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2171,16 +2186,58 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
+	if (sub) {
+		bt_addr_le_copy(&sub->peer, &conn->le.dst);
+		return sub;
+	}
```
The pattern I've used elsewhere for this, in order to avoid iterating the list twice, is to pass a second "free_slot" parameter to the lookup function. The lookup function will set that to point to the first empty slot in case the lookup of the actual connection fails: https://github.com/zephyrproject-rtos/zephyr/blob/8067aca7e90617fb49fb0895066c164d60559198/subsys/bluetooth/mesh/cfg_srv.c#L1107-L1131

Would something like that make sense here?

Vudentz replied:
To be honest I haven't consider this would be a big deal since MAX_CONN+MAX_PAIRED is normally not that big, but I can use the same logic to just return an unused slot in one go.

Vudentz replied:
Btw an alternative for this is to use ffs and then maintain a bitmap i.e:

```
/* Find unique id in range from 1 to max but no bigger then
 * sizeof(int) * 8. ffs() is used since it is POSIX standard
 */
uint8_t util_get_uid(unsigned int *bitmap, uint8_t max)
{
	uint8_t id;

	id = ffs(~*bitmap);

	if (!id || id > max)
		return 0;

	*bitmap |= 1 << (id - 1);

	return id;
}

/* Clear id bit in bitmap */
void util_clear_uid(unsigned int *bitmap, uint8_t id)
{
	if (!id)
		return;

	*bitmap &= ~(1 << (id - 1));
}
```

Vudentz committed:
Bluetooth: shell: Fix not reseting value_handle if subscription fails

If bt_gatt_subscirbe fails value_handle must be reset since otherwise it
will not possible to subscribe again as the parameters will be consider
in use.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@jhedberg Anything else that needs fixing before we merge this one?

wdhpawa reviewed (approved):
Looks OK to me.

joerchan commented:
@Vudentz Did you want to do the ffs bitmap thing? Because I think we could CLZ/CTZ as well for that (if  I understood correctly). The are implemented as:  u32_count_leading_zeros and u32_count_trailing_zeros.

jhedberg commented:
> @jhedberg Anything else that needs fixing before we merge this one?

@Vudentz I think @joerchan pointed out some valid issues, so I'd wait for their resolution first

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -85,7 +88,7 @@ struct bt_att {
 };
 
-static struct bt_att bt_req_pool[CONFIG_BT_MAX_CONN];
+static struct bt_att bt_req_slab[CONFIG_BT_MAX_CONN];
```
Im fixing it was unintented change.

Vudentz replied:
Done, i end up moving this to use k_mem_slab as well.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1602,13 +1607,23 @@ static int gatt_send(struct bt_conn *conn, struct net_buf *buf,
+
+		/* Allocate new request */
+		req = bt_att_req_alloc(K_FOREVER);
+		if (!req) {
```
It should probably be using a 30 seconds timeout instead like other parts of the code.

Vudentz committed:
Bluetooth: ATT: Use k_mem_slab for connection context

This uses k_mem_slab APIs to allocate/free ATT context instead of custom
array.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
